Enable PublishAot where available

eerhardt opened the pull request:
This enables native AOT publish process during the PR validation and official builds when it can (just win-x64 and win-arm64 on official builds for now). When we add more official build targets like macOS and linux, those will light up with this change.

We need to split the Aspire.Cli project into 2 separate projects because in order to publish for AOT, it is easiest to set `PublishAot=true` in the .csproj directly. Trying to pass it in during publish causes lots of problems.

Because of https://github.com/dotnet/runtime/issues/94178, this doesn't fail when there are warnings because PublishAot in net8.0 always emits warnings, which don't get turned into errors.

eerhardt committed:
Enable PublishAot in PR validation

This exercises the native AOT publish process during the PR validation.

Because of https://github.com/dotnet/runtime/issues/94178, this doesn't fail when there are warnings because PublishAot in net8.0 always emits warnings, which don't get turned into errors.

mitchdenny reviewed (approved):
Good idea to catch these issues early.

radical committed:
Build the per-os archives for cli with AOT

.. and use `dotnet publish ..` for publishing instead of just invoking
the `Publish` target.

radical committed:
Skip AOT on azdo

radical committed:
upload logs from build_packages step if it fails

radical committed:
fix yml

radical committed:
fix

eerhardt commented on eng/clipack/Common.projitems:
```diff
@@ -13,19 +13,14 @@
-        Targets="publish"
-        Properties="@(AdditionalProperties)"
-        RemoveProperties="OutputPath;TargetFramework" />
+    <Exec Command="dotnet publish $(RepoRoot)src/Aspire.Cli/Aspire.Cli.csproj @(AdditionalProperties -> '-p:%(Identity)', ' ')" />
```
There are outstanding issues on why `dotnet msbuild /t:Publish` doesn't act the same as `dotnet publish`. https://github.com/dotnet/sdk/issues/32272 describes the issues.

I haven't fully figured out all the issues here, but this script shows the problems:

```
❯ dotnet restore
Restore complete (0.2s)

Build succeeded in 0.4s
❯ dotnet msbuild /t:Publish /p:PublishAot=true
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  AotTest succeeded (1.6s) → bin\Debug\net10.0\publish\

Build succeeded in 1.8s
❯ dotnet publish /p:PublishAot=true
Restore complete (2.0s)
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  AotTest succeeded (3.6s) → bin\Release\net10.0\win-x64\publish\

Build succeeded in 5.8s
❯ dotnet clean
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy

Build succeeded in 0.3s
❯ dotnet restore /p:PublishAot=true
Restore complete (0.3s)

Build succeeded in 0.5s
❯ dotnet msbuild /t:Publish /p:PublishAot=true
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  AotTest failed with 1 error(s) (0.3s) → bin\Debug\net10.0\AotTest.dll
    D:\packages\NuGet\cache\microsoft.dotnet.ilcompiler\10.0.0-preview.5.25277.114\build\Microsoft.NETCore.Native.Publish.targets(50,5): error : RuntimeIdentifier is required for native compilation. Try running dotnet publish with the -r option value specified.

Build failed with 1 error(s) in 0.5s
❯ dotnet msbuild /t:Publish /p:PublishAot=true /p:RuntimeIdentifier=win-x64
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  AotTest failed with 1 error(s) (0.1s)
    C:\Program Files\dotnet\sdk\10.0.100-preview.5.25277.114\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1047: Assets file 'D:\DotNetTest\AotTest\obj\project.assets.json' doesn't have a target for 'net10.0/win-x64'. Ensure that restore has run and that you have included 'net10.0' in the TargetFrameworks for your project. You may also need to include 'win-x64' in your project's RuntimeIdentifiers.

Build failed with 1 error(s) in 0.3s
❯ dotnet restore /p:PublishAot=true /p:RuntimeIdentifier=win-x64
Restore complete (0.3s)

Build succeeded in 0.5s
❯ dotnet msbuild /t:Publish /p:PublishAot=true /p:RuntimeIdentifier=win-x64
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  AotTest failed with 1 error(s) (1.9s) → bin\Debug\net10.0\win-x64\AotTest.dll
    D:\packages\NuGet\cache\microsoft.net.illink.tasks\10.0.0-preview.5.25277.114\build\Microsoft.NET.ILLink.targets(209,5): error NETSDK1102: Optimizing assemblies for size is not supported for the selected publish configuration. Please ensure that you are publishing a self-contained app.
```

I think the problem is I need to specify `/p:_IsPublishing=true` on both Restore and Publish.

eerhardt replied:
I ended up trying to split the projects into 2 separate Cli projects. one for AOT, one for the global tool.

eerhardt committed:
Split the CLI global tool into a separate project

eerhardt committed:
Set PublishTrimmed to false in the Cli Tool because PublishAot turns it on.

eerhardt committed:
disable Native AOT on linux-arm64

eerhardt committed:
Move non-AOT settings into Common.projitems

eerhardt committed:
Always native AOT the CLI if running on the target platform.

eerhardt committed:
Move NuGet properties back into Aspire.Cli.csproj

eerhardt committed:
Remove the Xliff resource folders

radical commented:
Internal build - https://dev.azure.com/dnceng/internal/_build/results?buildId=2738175&view=results

radical commented:
Public azdo build with tests - https://dev.azure.com/dnceng-public/public/_build/results?buildId=1079356&view=results

eerhardt committed:
Use PackageId instead of project name in GetPackageName

radical commented:
New azdo build with tests - https://dev.azure.com/dnceng-public/public/_build/results?buildId=1079429&view=results

eerhardt committed:
Remove AOT symbols from the publish directory

radical commented:
Checks before merging:

- [ ] Confirm AOT publish ran for CLI from binlogs
- [ ] internal build: confirm the Cli tool nuget is produced correctly, and works
- [ ] internal build: check the Cli zip for `win-x64, win-arm64` has the AOT build, and works
- [ ] no related test failures on public azdo build

radical committed:
Fix command line to generate binlog

radical committed:
Fix playground tests build on helix

`AzureFunctionsEndToEnd.AppHost.csproj : error NU1010: The following PackageReference items do not define a corresponding PackageVersion item: Aspire.Hosting.Azure.AppContainers. Projects using Central Package Management must declare PackageReference and PackageVersion items with matching names. For more information, visit https://aka.ms/nuget/cpm/gettingstarted `

radical committed:
Fix helix test run for template tests

eerhardt committed:
Fix command line on linux to use dash instead of slash
